algorithms/src/std_algorithms/numeric/Kokkos_TransformReduce.hpp

```diff
@@ -0,0 +1,496 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 3.0
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_STD_NUMERICS_TRANSFORM_REDUCE_HPP
+#define KOKKOS_STD_NUMERICS_TRANSFORM_REDUCE_HPP
+
+#include <Kokkos_Core.hpp>
+#include "../Kokkos_Constraints.hpp"
+#include "../Kokkos_ModifyingOperations.hpp"
+#include "../Kokkos_BeginEnd.hpp"
+#include "../Kokkos_ReducerWithArbitraryJoinerNoNeutralElement.hpp"
+
+namespace Kokkos {
```

Comment:
Really minor, but if `transform_{inclusive|exclusive}_scan` is in the same header as `{inclusive|exclusive}_scan` shouldn't this be in the same header as `reduce`? Or alternatively have everything separate.

Adding on to this I think there could be a lot less duplicated code if they reduce and transform reduce used the same basic code, with the transform being identity for the regular reduce (of course testing to see if it was no-op'd out by various compilers)

Reply:
yes, good observations! we have not yet fully finalized the organization. Will take a look at this. 
In most cases, we explicitly separated functors with default predicates from those accepting custom ones to avoid unnecessary layers even if compiler *might* optimize that. I can dig a bit deeper into that. But we know for sure that the current impl as is avoid those layers.